### ISSUE TITLE
calling as_ptr from `__asan_default_options` segfaults with `-C debuginfo=2`

### ISSUE THREAD

**Issue by Hello71** [rust-lang/rust#50066]
in the process of working around #41807, I found that the following code fragment results in a segmentation fault when run after compilation with `-Z sanitizer=address -C debuginfo=2`.

```rust
const ASAN_DEFAULT_OPTIONS: &'static [u8] = b"detect_odr_violation=1\0";

#[no_mangle]
pub extern "C" fn __asan_default_options() -> *const u8 {
    ASAN_DEFAULT_OPTIONS.as_ptr()
}

fn main() {
    println!("Hello World!");
}
```

```
$ rustc test.rs -Z sanitizer=address
$ ./test
Hello World!
$ rustc test.rs -Z sanitizer=address -C debuginfo=2
$ ./test
sh: segmentation fault (core dumped)  ./test
$ rust-gdb ./test
[ ... ]
Program received signal SIGSEGV, Segmentation fault.
0x000055555555e11d in alloc::slice::<impl [T]>::as_ptr (self=&[u8](len: 140737488345840) = {...})
    at /checkout/src/liballoc/slice.rs:474
(gdb) bt
#0  0x000055555555e11d in alloc::slice::<impl [T]>::as_ptr (self=&[u8](len: 140737488345840) = {...})
    at /checkout/src/liballoc/slice.rs:474
#1  0x000055555555e2ae in __asan_default_options () at test.rs:5
#2  0x000055555556da92 in __asan::MaybeCallAsanDefaultOptions ()
    at /checkout/src/libcompiler_builtins/compiler-rt/lib/asan/asan_flags.cc:31
#3  __asan::InitializeFlags () at /checkout/src/libcompiler_builtins/compiler-rt/lib/asan/asan_flags.cc:115
#4  0x000055555555cd25 in __asan::AsanInitInternal ()
    at /checkout/src/libcompiler_builtins/compiler-rt/lib/asan/asan_rtl.cc:380
#5  0x00007ffff7de75de in _dl_init () from /lib64/ld-linux-x86-64.so.2
#6  0x00007ffff7dd8f6a in _dl_start_user () from /lib64/ld-linux-x86-64.so.2
#7  0x0000000000000001 in ?? ()
#8  0x00007fffffffe062 in ?? ()
#9  0x0000000000000000 in ?? ()
```

Interestingly, adding `-C opt-level=1` fixes the issue, as does manually inlining `as_ptr` (i.e. `ASAN_DEFAULT_OPTIONS as *const [u8] as *const u8`).

**Comment by tmiasko** [comment]
The address sanitizer needs to complete initialization before calling
instrumented functions is possible. 

The `__asan_default_options` itself is never instrumented (the functions with
names starting with `__asan_` are excluded from instrumentation), but it cannot
call other functions that might be, like `as_ptr` for example.

**Comment by Hello71** [comment]
It'd be nice if it didn't crash though. Also, isn't it still a problem that it's contingent on `-C opt-level=1` and possibly also `-C debuginfo=2`? What's the point of instrumenting `as_ptr` if it gets inlined anyways?

**Comment by tmiasko** [comment]
> What's the point of instrumenting as_ptr if it gets inlined anyways?

It isn't inlined actually when the issue occurs.

> It'd be nice if it didn't crash though.

You could try reporting issue upstream to google/sanitizers or bugs.llvm.org,
since I don't think there is much that could be done on rustc side of things,
nor is the issue specific to rustc.

**Comment by Mark-Simulacrum** [comment]
I'm going to close this as expected, if perhaps undesirable, behavior, and changes here being out of scope for rustc.
